src/jsdoc_transformer.ts

```diff
@@ -0,0 +1,937 @@
+          tagName: '',
+          text: 'NOTE: tsickle could not resolve supertype, ' +
+              'class definition may be incomplete.\n'
+        });
+        continue;
+      }
+      let alias: ts.Symbol = sym;
+      if (sym.flags & ts.SymbolFlags.TypeAlias) {
+        // It's implementing a type alias.  Follow the type alias back
+        // to the original symbol to check whether it's a type or a value.
+        const type = mtt.typeChecker.getDeclaredTypeOfSymbol(sym);
+        if (!type.symbol) {
+          // It's not clear when this can happen, but if it does all we
+          // do is fail to emit the @implements, which isn't so harmful.
+          continue;
+        }
+        alias = type.symbol;
+      }
+      if (alias.flags & ts.SymbolFlags.Alias) {
+        alias = mtt.typeChecker.getAliasedSymbol(alias);
+      }
+      const typeTranslator = mtt.newTypeTranslator(impl.expression);
+      if (typeTranslator.isBlackListed(alias)) {
+        continue;
+      }
+      // We can only @implements an interface, not a class.
+      // But it's fine to translate TS "implements Class" into Closure
+      // "@extends {Class}" because this is just a type hint.
+      if (alias.flags & ts.SymbolFlags.Class) {
+        if (!isClass) {
+          // Only classes can extend classes in TS. Ignoring the heritage clause should be safe,
+          // as interfaces are @record anyway, so should prevent property disambiguation.
+
+          // Problem: validate that methods are there?
+          continue;
+        }
+        if (classHasSuperClass && heritage.token !== ts.SyntaxKind.ExtendsKeyword) {
+          // Do not emit an @extends for a class that already has a proper ES6 extends class. This
+          // risks incorrect optimization, as @extends takes precedence, and Closure won't be
+          // aware of the actual type hierarchy of the class.
+          continue;
+        }
+        tagName = 'extends';
+      } else if (alias.flags & ts.SymbolFlags.Value) {
+        // If the symbol was already in the value namespace, then it will
+        // not be a type in the Closure output (because Closure collapses
+        // the type and value namespaces).  Just ignore the implements.
+        continue;
+      }
+      // typeToClosure includes nullability modifiers, so call symbolToString directly here.
+      docTags.push({tagName, type: typeTranslator.symbolToString(alias, true)});
+    }
+  }
+}
+
+/** Flags that declare a field of the same name if set on a ctor parameter. */
+const FIELD_DECLARATION_MODIFIERS: ts.ModifierFlags = ts.ModifierFlags.Private |
+    ts.ModifierFlags.Protected | ts.ModifierFlags.Public | ts.ModifierFlags.Readonly;
+
+/**
+ * emitMemberTypes emits the type annotations for members of a class. It's necessary in the case
+ * where TypeScript syntax specifies there are additional properties on the class, because to
+ * declare these in Closure you must declare these separately from the class.
+ */
```

Comment:
Doc the return type here, because it's a bit funny?

Reply:
Done.